src/models/transfers.js

```diff
@@ -0,0 +1,354 @@
+'use strict'
+
+const crypto = require('crypto')
+const _ = require('lodash')
+const diff = require('deep-diff')
+const tweetnacl = require('tweetnacl')
+const stringifyJSON = require('canonical-json')
+const db = require('../services/db')
+const makeAccountBalances = require('../lib/accountBalances')
+const validateNoDisabledAccounts = require('../lib/disabledAccounts')
+const config = require('../services/config')
+const uri = require('../services/uriManager')
+const transferExpiryMonitor = require('../services/transferExpiryMonitor')
+const notificationWorker = require('../services/notificationWorker')
+const log = require('../services/log')('transfers')
+const requestUtil = require('five-bells-shared/utils/request')
+const updateState = require('../lib/updateState')
+const hashJSON = require('five-bells-shared/utils/hashJson')
+const Transfer = require('./db/transfer').Transfer
+const NotFoundError = require('five-bells-shared/errors/not-found-error')
+const InvalidUriParameterError = require('five-bells-shared/errors/invalid-uri-parameter-error')
+const UnmetConditionError = require('five-bells-shared/errors/unmet-condition-error')
+const InvalidModificationError =
+require('five-bells-shared/errors/invalid-modification-error')
+const UnprocessableEntityError =
+require('five-bells-shared/errors/unprocessable-entity-error')
+const UnauthorizedError =
+require('five-bells-shared/errors/unauthorized-error')
+
+function * getTransfer (id) {
+  log.debug('fetching transfer ID ' + id)
+
+  let transfer = yield Transfer.findById(id)
+  if (!transfer) {
+    throw new NotFoundError('Unknown transfer ID')
+  }
+
+  return transfer.getDataExternal()
+}
+
+function * getTransferStateReceipt (id, signatureType, conditionState) {
+  log.debug('fetching state receipt for transfer ID ' + id)
+  const transfer = yield Transfer.findById(id)
+  const transferState = transfer ? transfer.state : 'nonexistent'
+  const message = {
+    id: uri.make('transfer', id),
+    state: transferState
+  }
+  const transferStateReceipt = {
+    type: signatureType,
+    message: message,
+    signer: config.server.base_uri
+  }
+
+  if (signatureType === 'ed25519-sha512') {
+    transferStateReceipt.public_key = config.keys.ed25519.public
+    transferStateReceipt.signature = sign(hashJSON(message))
+  } else if (signatureType === 'sha256') {
+    const realPreImage = makePreImage(message.id, transferState)
+    transferStateReceipt.digest = sha256(stringifyJSON(realPreImage))
+    transferStateReceipt.message = makePreImage(message.id, transferState)
+    if (conditionState) {
+      const conditionPreImage = makePreImage(message.id, conditionState)
+      transferStateReceipt.condition_state = conditionState
+      transferStateReceipt.condition_digest = sha256(stringifyJSON(conditionPreImage))
+    }
+  } else {
+    // TODO: move this validation to controller or remove "URI"
+    throw new InvalidUriParameterError('type is not valid')
```

Comment:
I think this can go in the controller. Provided we only support a few algorithms, we can just validate that the URI param is one of them. If the list grows, we could check against some kind of system dictionary.